### ISSUE TITLE
objectiveとeval_metricで共通のmetricが重複計算される

### ISSUE THREAD

**Issue by matsuken92** [matsuken92/LightGBM_test#1]
objective="regression_l1"
'eval_metric': ['l2', 'l1']
のように、objectiveに紐づくmetircが付与され、かつそれがeval_metricで同じmetricがあった時に重複して計算されてしまう。

**Comment by matsuken92** [comment]
#### Test Pattern
```
        X, y = load_boston(True)
        X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.1, random_state=42)

        eval_result = {}
        callbacks = [lgb.record_evaluation(eval_result)]

        params_fit = {'X': X_train,
                      'y': y_train,
                      'eval_set': (X_test, y_test),
                      'early_stopping_rounds': 5,
                      'verbose': False,
                      'eval_metric': ['l2', 'l1']}
        model = lgb.LGBMRegressor(objective="regression_l1", n_estimators=10)
        gbm = model.fit(**params_fit, callbacks=callbacks)
        print(gbm.evals_result_)
```

env.params["metric"]には `[l2, l1, regression_l1]`  が入っている。
![スクリーンショット 2019-06-26 22 40 19](https://user-images.githubusercontent.com/7313648/60184732-890c9c00-9863-11e9-86b0-4aad2c412273.png)


callback.pyのenv.evaluation_result_listには `[l1, l2, l1]` が入っている。
![スクリーンショット 2019-06-26 22 37 41](https://user-images.githubusercontent.com/7313648/60184552-37641180-9863-11e9-9fdd-71ea0f009316.png)

engine.py l.254 周辺 evaluation_result_listに値を詰めているところ
```py
        evaluation_result_list = []
        # check evaluation result.
        if valid_sets is not None:
            if is_valid_contain_train:
                evaluation_result_list.extend(booster.eval_train(feval))
            evaluation_result_list.extend(booster.eval_valid(feval))
        try:
            for cb in callbacks_after_iter:
                cb(callback.CallbackEnv(model=booster,
                                        params=params,
                                        iteration=i,
                                        begin_iteration=init_iteration,
                                        end_iteration=init_iteration + num_boost_round,
                                        evaluation_result_list=evaluation_result_list))
```

basic.py l. 2107周辺 `booster.eval_valid()` 関数
```py

        return [item for i in range_(1, self.__num_dataset)
                for item in self.__inner_eval(self.name_valid_sets[i - 1], i, feval)]
```

basic.py l.2552  `__inner_eval` 関数
```py
    def __inner_eval(self, data_name, data_idx, feval=None):
        """Evaluate training or validation data."""
        if data_idx >= self.__num_dataset:
            raise ValueError("Data_idx should be smaller than number of dataset")
        self.__get_eval_info()
        ret = []
        if self.__num_inner_eval > 0:
            result = np.zeros(self.__num_inner_eval, dtype=np.float64)
            tmp_out_len = ctypes.c_int(0)
            _safe_call(_LIB.LGBM_BoosterGetEval(
                self.handle,
                ctypes.c_int(data_idx),
                ctypes.byref(tmp_out_len),
                result.ctypes.data_as(ctypes.POINTER(ctypes.c_double))))
            if tmp_out_len.value != self.__num_inner_eval:
                raise ValueError("Wrong length of eval results")
            for i in range_(self.__num_inner_eval):
                ret.append((data_name, self.__name_inner_eval[i], # ★★★ self.__name_inner_evalは[l1, l2, l1]になっていてenv.params["metric"]と順番が違う
                            result[i], self.__higher_better_inner_eval[i]))
        if feval is not None:
            if data_idx == 0:
                cur_data = self.train_set
            else:
                cur_data = self.valid_sets[data_idx - 1]
            feval_ret = feval(self.__inner_predict(data_idx), cur_data)
            if isinstance(feval_ret, list):
                for eval_name, val, is_higher_better in feval_ret:
                    ret.append((data_name, eval_name, val, is_higher_better))
            else:
                eval_name, val, is_higher_better = feval_ret
                ret.append((data_name, eval_name, val, is_higher_better))
        return ret
```


booster = Booster(params=params, train_set=train_set)

basic.py l.2630 `__get_eval_info()` 関数内 
`booster = Booster(params=params, train_set=train_set)` が呼ばれた時に呼ばれる
```py
                self.__name_inner_eval = \
                    [string_buffers[i].value.decode() for i in range_(self.__num_inner_eval)]
```

`LGBM_BoosterGetEvalNames` で順番がおかしくなってるっぽい


### 解決法
処理済みのmetricsは2回目以降追加しない
```py

        processed_metric = []
        for data_name, eval_name, result, _ in env.evaluation_result_list:
            if eval_name in processed_metric: continue
            if eval_name not in eval_result[data_name].keys():
                eval_result[data_name][eval_name] = []
            eval_result[data_name][eval_name].append(result)
            processed_metric.append(eval_name)
```

**Comment by matsuken92** [comment]
#### Test Pattern2

```py

        params = {'n_estimators': 25,
                  'learning_rate': 0.8,
                  'num_leaves': 15,
                  'verbose': -1,
                  'seed': 123}
        params_fit = {'X': X_train,
                      'y': y_train,
                      'eval_set': (X_train, y_train),
                      'early_stopping_rounds': 5,
                      'verbose': False}
```

指定しないときはErrorにする？

**Comment by matsuken92** [comment]
### Aliases for metrics
https://lightgbm.readthedocs.io/en/latest/Parameters.html#metric

```
"l1" ["mean_absolute_error", "mae", "regression_l1"]
"l2" ["mean_squared_error", "mse", "regression_l2", "regression"]
"l2_root" ["root_mean_squared_error", "rmse"]
"mape" ["mean_absolute_percentage_error", ]
"ndcg" ["lambdarank", ]
"map" ["mean_average_precision", ]
"binary_logloss" ["binary", ]
"multi_logloss" ["multiclass", "softmax", "multiclassova", "multiclass_ova", "ova", "ovr" ]
"xentropy" ["cross_entropy", ]
"xentlambda" ["cross_entropy_lambda", ]
"kldiv" ["kullback_leibler", ]
```
